common/src/main/java/io/netty/util/ResourceLeakDetector.java

```diff
@@ -343,42 +340,29 @@ protected void reportInstancesLeak(String resourceType) {
             // It's important that we not store a reference to the referent as this would disallow it from
             // be collected via the PhantomReference.
             trackedHash = System.identityHashCode(referent);
-
-            Level level = getLevel();
-            if (level.ordinal() >= Level.ADVANCED.ordinal()) {
-                creationRecord = newRecord(null, 3);
-                lastRecords = new ArrayDeque<String>();
-            } else {
-                creationRecord = null;
-                lastRecords = null;
-            }
+            head = tail = getLevel().ordinal() >= Level.ADVANCED.ordinal() ? new Record() : null;
             allLeaks.put(this, LeakEntry.INSTANCE);
         }
 
         @Override
         public void record() {
-            record0(null, 3);
+            record0(null);
         }
 
         @Override
         public void record(Object hint) {
-            record0(hint, 3);
+            record0(hint);
         }
 
-        private void record0(Object hint, int recordsToSkip) {
+        private void record0(Object hint) {
             // Check MAX_RECORDS > 0 here to avoid similar check before remove from and add to lastRecords
-            if (creationRecord != null && MAX_RECORDS > 0) {
-                String value = newRecord(hint, recordsToSkip);
-
-                synchronized (lastRecords) {
-                    int size = lastRecords.size();
-                    if (size == 0 || !lastRecords.getLast().equals(value)) {
-                        if (size >= MAX_RECORDS) {
-                            lastRecords.removeFirst();
-                            ++removedRecords;
-                        }
-                        lastRecords.add(value);
-                    }
+            if (head != null && MAX_RECORDS > 0) {
+                Record record = hint == null ? new Record() : new Record(hint);
+
+                synchronized (head) {
+                    tail.next = record;
+                    tail = record;
+                    numRecords++;
```

Comment:
Since I can't comment on the actual line:

On line 373 where you call close.  It looks like you should call clear() too, which will avoid the ref being added to the queue at all.

Reply:
good idea!